Add functionality for components in resource pack descriptions.

cullvox opened the pull request:
Attempt to Fix #1602

Allows for pack.mcmeta file descriptions that are written using components to now be readable while still allowing for normal strings. 

I tested this fix on Windows, and it reads in as expected. 

Colors are not implemented. They would either have to be converted to original color codes, or a more advanced solution would be required.

cullvox committed:
added support for components in resource pack descriptions.

Signed-off-by: Caden Miller <me@cadenmiller.dev>

cullvox committed:
fix style, and use qWarning instead of throw.

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
DCO Remediation Commit for cullvox <cullvox@outlook.com>

I, cullvox <cullvox@outlook.com>, hereby add my Signed-off-by to this commit: d2e662ddbb518b147de76ca3f613e046d2912db5

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
fix clang-format failing

cullvox committed:
DCO Remediation Commit for cullvox <cullvox@outlook.com>

I, cullvox <cullvox@outlook.com>, hereby add my Signed-off-by to this commit: 05f4214cc5d3a9005c0d259ca185303792aa2fa7

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
remove space for clang-format

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
fix not properly reading json text.

The text now displays properly in the GUI of Prism.

Signed-off-by: cullvox <cullvox@outlook.com>

Trial97 commented:
Well I'm ok with your attempt but please note that the document mentioned here:https://minecraft.fandom.com/wiki/Raw_JSON_text_format suggest that the format is recursive. 
Should this PR only implement the reading of the mentioned format? Or do you have plans to also correctly display the description with formatting ?

cullvox commented:
Thank you for showing me the actual documentation for this. I have looked it up many times and never found it.
I plan on doing what it takes to get this properly working. However, I can't say that I understand the code base very well yet, so it might take some time. I believe I understand JSON parsing in this project well enough to get the recursive format for text.

What I'm not sure about how colors would be properly implemented. My quick solution would be to convert these new components into a normal string and colors/format to chat codes. It's the only solution I can think of that doesn't require changing some other file that I don't know exists yet, or making a new one.

Trial97 commented:
Regarding the text color/format my suggestion is just to convert them to HTML elements.
For example, if you have a JSON element like this: `{"text": "some cool  colored text", "color":"dark_blue"}` you would translate it as  the following string: `<span style="color: dark_blue"};">some cool  colored text</span>` (the QLabel can handle HTML) 
Obviously, you can nest tags and stuff as needed.
Another suggestion is to do all this decoding in a separate function so you can reuse it if needed.

cullvox commented:
Ok nice thanks for that suggestion, makes it easier knowing that QLabel will handle HTML! Yeah, I knew that once I saw that this problem was a little more complex than I thought, it would need to be split up into more functions.

cullvox committed:
added formatting with colors.

This is somewhat dirty implementation and I will clean it up soon.
Using the HTML rich text features of Qt, made it much easier to
understand what was needed.

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
fix formatting and add more proper errors.

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
attempt to fix clang-format and ubuntu build.

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox commented:
Formatting is a little better now instead of just boring text we support the colors and formats.

Json like this:
```JSON {
{
  "pack": {
    "pack_format": 15,
    "description": [
      "",
      {
        "text": "Welcome to ",
        "bold": true,
        "extra": [
          {
            "text": "THIS IS SOME EXTRA",
            "italic": true
          }
        ]
      },
      {
        "text": "Minecraft\n",
        "bold": true,
        "color": "#A0C5D2"
      },
      {
        "text": " Tools ",
        "bold": true
      },
      {
        "text": "some t"
      },
      {
        "text": "ext ",
        "color": "white",
        "italic": true
      },
      {
        "text": "SomeText",
        "bold": true,
        "italic": true,
        "color": "gold"
      },
      {
        "text": " TEST",
        "bold": true,
        "italic": true,
        "obfuscated": true,
        "color": "aqua"
      }
    ]
  }
}
```

Now presents HTML like this:
```HTML
<font color=\"#000000\"><b>Welcome to <font color=\"#000000\"><i>THIS IS SOME EXTRA</font></i></font></b>
<font color=\"#A0C5D2\"><b>Minecraft\n</font></b>
<font color=\"#000000\"><b> Tools </font></b>
<font color=\"#000000\">some t</font>
<font color=\"#FFFFFF\"><i>ext </font></i>
<font color=\"#FFAA00\"><b><i>SomeText</font></b></i>
<font color=\"#55FFFF\"><b><i> TEST</font></b></i>
```

This seems to pass most of the tests required. If there is something else that I have missed, I'd like to know. The only thing I can think about is obscured text, and I don't think HTML will save me there, so maybe it's best just to leave it. I have to do some more testing on strike though and underline, but I think that they should work too.

EDIT: Also working on getting cland-format and the other check to work.

Trial97 commented:
I'm okay with `obfuscated` not being handled, but can you look into `clickEvent` and more specifically `open_url` that can be translated into `<a href="URL">...</a>`.
Regarding the mentioned tests can you also add tests for this implementation, or it's a little too much?
For clang checks, you can use the following command: `clang-format --style="file" -i path/to/file`.

Trial97 commented on launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp:
```diff
@@ -178,6 +179,168 @@ bool processZIP(ResourcePack& pack, ProcessingLevel level)
+    bool strikethrough = false;
+};
+
+bool textColorToHexColor(const QString& text_color, QString& result)
+{
+    static const QHash<QString, QString> text_to_hex = {
+        { "black", "#000000" },     { "dark_blue", "#0000AA" },    { "dark_green", "#00AA00" }, { "dark_aqua", "#00AAAA" },
+        { "dark_red", "#AA0000" },  { "dark_purple", "#AA00AA" },  { "gold", "#FFAA00" },       { "gray", "#AAAAAA" },
+        { "dark_gray", "#555555" }, { "blue", "#5555FF" },         { "green", "#55FF55" },      { "aqua", "#55FFFF" },
+        { "red", "#FF5555" },       { "light_purple", "#FF55FF" }, { "yellow", "#FFFF55" },     { "white", "#FFFFFF" },
+    };
+
+    auto it = text_to_hex.find(text_color);
+
+    if (it == text_to_hex.end()) {
+        qWarning() << "Invalid color string in component!";
+        result = "#000000"; // return black if color not found
+        return false;
+    }
+
+    result = it.value();
+    return true; 
+}
```
I do believe HTML can handle the color names and you do not need to do any conversion.

Trial97 commented on launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp:
```diff
@@ -186,7 +349,21 @@ bool processMCMeta(ResourcePack& pack, QByteArray&& raw_data)
+        // description could either be string, or array of dictionaries
+        auto desc_val = pack_obj.value("description");
+
+        if (desc_val.isString()) {
+            pack.setDescription(desc_val.toString());
```
The `desc_val` can also be any JSON type: 
```
A raw JSON text component can be any JSON data type other than JSON null, including JSON string, JSON array, JSON object, JSON boolean, and JSON number
```
so you would need to also handle null,object, boolean, and numbers.

Trial97 commented on launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp:
```diff
@@ -178,6 +179,168 @@ bool processZIP(ResourcePack& pack, ProcessingLevel level)
+
+void appendEndFormat(TextFormat& format, QString& toAppend)
+{
+    toAppend.append("</font>");
+    if (format.bold)
+        toAppend.append("</b>");
+    if (format.italic)
+        toAppend.append("</i>");
+    if (format.underlined)
+        toAppend.append("</u>");
+    if (format.strikethrough)
+        toAppend.append("</s>");
```
If you use multiple tags the end tags should be in reverse order of the beginning so instead of getting:
```<font color=\"#FFAA00\"><b><i>SomeText</font></b></i>```
you would get: 
```<font color=\"#FFAA00\"><b><i>SomeText</i></b></font>```
That is more correct in HTML

Trial97 commented on launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp:
```diff
@@ -178,6 +179,168 @@ bool processZIP(ResourcePack& pack, ProcessingLevel level)
+    } else if (value.isObject()) {
+        auto obj = value.toObject();
+
+        TextFormat format{};
+        if (parentFormat)
+            format = *parentFormat;
```
Because you are translating things in HTML you do not need to pass the parentFormat to the next function.
You already surrounded the text with bold or italic you do not need to do that anymore.

Trial97 commented on launcher/minecraft/mod/tasks/LocalResourcePackParseTask.cpp:
```diff
@@ -178,6 +179,168 @@ bool processZIP(ResourcePack& pack, ProcessingLevel level)
+    }
+}
+
+bool readFormat(const QJsonObject& obj, TextFormat& format)
+{
+    auto text = obj.value("text");
+    auto color = obj.value("color");
+    auto bold = obj.value("bold");
+    auto italic = obj.value("italic");
+    auto underlined = obj.value("underlined");
+    auto strikethrough = obj.value("strikethrough");
+    auto extra = obj.value("extra");
+
+    if (color.isString()) {
+        // colors can either be a hex code or one of a few text colors
+        auto col_str = color.toString();
+
+        const QRegularExpression hex_expression("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$");
+        const auto hex_match = hex_expression.match(col_str);
+
+        if (hex_match.hasMatch()) {
+            format.color = color.toString();
+        } else {
+            if (!textColorToHexColor(color.toString(), format.color)) {
+                qWarning() << "Invalid color type in component!";
+                return false;
+            }
+        }
+    }
+
+    return getBoolOrFromText(bold, format.bold) && getBoolOrFromText(italic, format.italic) &&
+           getBoolOrFromText(underlined, format.underlined) && getBoolOrFromText(strikethrough, format.strikethrough);
+}
```
```suggestion
bool readFormat(const QJsonObject& obj, TextFormat& format)
{
    format.color = Json::ensureString(obj, "color");
    format.bold = Json::ensureBoolean(obj, "bold", false);
    format.italic = Json::ensureBoolean(obj, "italic", false);
    format.underlined = Json::ensureBoolean(obj, "underlined", false);
    format.strikethrough = Json::ensureBoolean(obj, "strikethrough", false);
    return true;
}
```

cullvox commented:
> I'm okay with `obfuscated` not being handled, but can you look into `clickEvent` and more specifically `open_url` that can be translated into `<a href="URL">...</a>`.

I didn't even think about it, but that's a great idea! I'll add that, along with updating with the review changes.

cullvox commented:
>Regarding the mentioned tests can you also add tests for this implementation, or it's a little too much?`

I can write tests, just not sure what process you want me to go about doing that. I looked up that Qt has a testing framework, and I'm thinking that's what I should do.

Trial97 commented:
Regarding tests, you can inspire from the existing ones:  https://github.com/PrismLauncher/PrismLauncher/tree/develop/tests
Create a new file in that folder and add the file to build in Cmake: https://github.com/PrismLauncher/PrismLauncher/blob/develop/tests/CMakeLists.txt like so:
```
ecm_add_test(NewFile_test.cpp LINK_LIBRARIES Launcher_logic Qt${QT_VERSION_MAJOR}::Test
    TEST_NAME PackDescription)
```

cullvox commented:
Currently, all the code for this is within the cpp file of "LocalResourcePackParseTask.cpp", if It's going to be used as a test (and possibly else where) should I make it public in the header of current cpp file it's in.  Alternatively, I could create a new file somewhere and make it public like that. It would be a little more generalized because if this project gets more in the data pack area, they also use components.

Trial97 commented:
Just make it public in the header file and use that in your tests.

cullvox committed:
clean up and add review suggestions, links open

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
added tests, fixed issues with overriding/format

In the documentation it states that child values can override the parent
values. Originally this code did not support that but now it does. Also
added in testing inspired by the previous tests.

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox committed:
clang-format fixes

cullvox committed:
DCO Remediation Commit for cullvox <cullvox@outlook.com>

I, cullvox <cullvox@outlook.com>, hereby add my Signed-off-by to this commit: ddf0c28b1b0b6617b0308405b4d5eefd20ff48b2

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox commented:
Not sure why check pre-commit failed, but everything else looks much better now. There was a change in formatting that made the suggestion about HTML and parenting kinda an issue. Using HTML, children cannot override something like `<u>` or `<b>` so it can't be disabled, so I had to change it up a little bit. 

There was a lot that changed in these commits, but I think for the better.

Trial97 commented:
The pre-commit check fails because you added some JSON files, they need to be formatted with `clang-format --style="file" -i path/to/file`.

cullvox committed:
use clang-format

Signed-off-by: cullvox <cullvox@outlook.com

cullvox committed:
DCO Remediation Commit for cullvox <cullvox@outlook.com>

I, cullvox <cullvox@outlook.com>, hereby add my Signed-off-by to this commit: ee7016fa542e3ca92bdd691bd2e168d9d43a02c2

Signed-off-by: cullvox <cullvox@outlook.com>

Trial97 commented:
This almost works: the parsing works(maybe there is a way to be simplified) but now when the HTML string(resulting from this) has more than 290 characters, the hide functionality(from InfoFrame) doesn't work; the `...` at the end are no longer a link because of HTML errors(we do not expect HTML there).
Can you look into it?

cullvox committed:
fixes html elide issue with InfoFrame

cullvox committed:
Merge branch 'components-resource-pack-fix' of https://github.com/cadenmiller/PrismLauncher into components-resource-pack-fix

cullvox committed:
DCO Remediation Commit for cullvox <cullvox@outlook.com>

I, cullvox <cullvox@outlook.com>, hereby add my Signed-off-by to this commit: c81689d39322d5ea5c7ed0073d73e4d8d2446b33

Signed-off-by: cullvox <cullvox@outlook.com>

cullvox commented:
This latest commits fix the HTML issue above and seems to work as expected now.

Trial97 committed:
simplify the raw json parsing

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Fixed Tests

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
removed missed header

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

cullvox committed:
Merge branch 'develop' into components-resource-pack-fix

Signed-off-by: cullvox <68567525+cadenmiller@users.noreply.github.com>
